includes/abilities-api/class-wp-abilities-registry.php

```diff
@@ -159,17 +147,23 @@ public function register( $name, array $properties = array() ): ?WP_Ability {
 			return null;
 		}
 
-		$ability = new WP_Ability(
+		/**
+		 * Filters the class used to instantiate the ability.
+		 *
+		 * This is helpful for creating complex ability types that are not well expressed by array values and callbacks.
+		 *
+		 * @param string $ability_class           The class name to instantiate the ability. Must extend \WP_Ability.
+		 * @param string $name                    The name of the ability being registered.
+		 * @param array<string,mixed> $properties The properties of the ability being registered. See
+		 *                                        `wp_register_ability()` for the expected structure.
+		 *
+		 * @phpstan-param class-string<\WP_Ability> $ability_class
+		 */
+		$ability_class = apply_filters( 'wp_ability_class', WP_Ability::class, $name, $properties );
```

Comment:
What do you think about making it easier use with the following?
```suggestion
		$ability_class = WP_Ability::class;
		if ( isset( $properties['ability_class'] ) } {
				$ability_class = $properties['ability_class'];
				unset( $properties['ability_class'] );
		}
		$ability_class = apply_filters( 'wp_ability_class', $ability_class, $name, $properties );
```

Technically speaking, this could work without a filter, too.

Reply:
Happy to go with this approach instead 🙇

> Technically speaking, this could work without a filter, too.

Not just "could",  I think for v0.1.0 (maybe even WP6.9 merge) we can remove it in favor of `$properties['ability_class']`, and then add the filter down the line (if still necessary and not addressed by e.g the filters in #38 or something else).

Reply:
I did a quick search and I see that [`register_post_type()`](https://developer.wordpress.org/reference/functions/register_post_type/) offers a way to set custom class names with some properties:

<img width="695" height="268" alt="Screenshot 2025-08-25 at 16 52 28" src="https://github.com/user-attachments/assets/f14148a8-4c65-4806-b7a4-14648ec1d3f0" />

I like this approach even more now.

I also noticed that some light validation would be appreciated when devs can modify the class:

https://github.com/WordPress/wordpress-develop/blob/1ab39ce269a0dd718f6e547162403c5eb11219ba/src/wp-includes/class-wp-post-type.php#L877-L893

In this case, it's probably fine to check if class exists and the instance is `WP_Ability` or extends it.

Reply:
Done in https://github.com/WordPress/abilities-api/pull/21/commits/d40a6f5f91c0f5f4a039d96b5e85049c4138e3e5 🚀